src/rtps/UART_node.cxx

```diff
@@ -0,0 +1,318 @@
+int UART_node::parseMavlinkFromUART(char buffer[], size_t buflen)
+{
+    unsigned i;
+
+    if (m_uart_filestream == -1)
+        return 2;
+
+    if (_buf_size < 3)
+        return 0;
+
+    // Search for a mavlink packet on buffer to send it
+    i = 0;
+    while (i < (_buf_size - 3) && _buf[i] != 253 && _buf[i] != 254)
+        i++;
+
+    // We need at least the first three bytes to get packet len
+    if (i == _buf_size - 3) {
+        return 0;
+    }
+
+    uint16_t packet_len;
+    if (_buf[i] == 253) {
+        uint8_t payload_len = _buf[i + 1];
+        uint8_t incompat_flags = _buf[i + 2];
+        packet_len = payload_len + 12;
+
+        if (incompat_flags & 0x1) { // signing
+            packet_len += 13;
+        }
+    } else {
+        packet_len = _buf[i + 1] + 8;
+    }
+
+    // cerr << "mavlink parse. message len " << (int)packet_len << " on: " << i << endl;
+
+    // packet is bigger than what we've read, better luck next time
+    if (i + packet_len > _buf_size) {
+        return 0;
+    }
+
+    // buffer should be big enough to hold a mavlink packet
+    if (packet_len > buflen) {
+        return -EMSGSIZE;
+    }
+
+    // Found a whole message, send it and remove from local _buf
+    memmove(buffer, _buf + i, packet_len);
+    memmove(_buf + i, _buf + i + packet_len, sizeof(_buf) - i - packet_len);
+    _buf_size -= packet_len;
+
+    return packet_len;
+}
+
+int UART_node::parseRTPSfromUART(char *topic_ID, uint8_t *seq, char buffer[], size_t buflen)
+{
+    size_t start = 0;
+    uint16_t len;
+    if (m_uart_filestream == -1)
+        return 2;
+
+    if (_buf_size < sizeof(struct Header)) // starting ">>>" + topic + seq + len + crchi + crclow
+        return 0;
+
+    // look for starting ">>>"
+    while (start <= (_buf_size - sizeof(struct Header)) && memcmp(_buf + start, ">>>", 3) != 0) {
+        start++;
+    }
+    if (start >= (_buf_size - sizeof(struct Header))) {
+        return 0;
+    }
+
+    struct Header *header = (struct Header*)&_buf[start];
+    len = ((uint16_t)header->payload_len_h << 8) | header->payload_len_l;
+    if (start + len > _buf_size)
+        return 0; // we don't have a complete msg yet
+
+    // cerr << "rtps parse. message len " << (int)len << " on: " << start << endl;
+
+    // buffer should be big enough to hold a rtps packet
+    if (len > buflen) {
+        return -EMSGSIZE;
+    }
+
+    // Found a whole message, send it and remove from local _buf
+    *topic_ID = header->topic_ID;
+    *seq = header->seq;
+
+    uint16_t crc = crc16(_buf + start + sizeof(struct Header), len);
+    uint16_t read_crc = ((uint16_t)header->crc_h << 8) | header->crc_l;
+    if (crc != read_crc)
+        log_debug("CRC mismatch %u - %u (%u) %d", crc, read_crc, len, header->topic_ID);
+    else
+        memmove(buffer, _buf + start + sizeof(struct Header), len);
+
+    memmove(_buf + start, _buf + start + sizeof(struct Header) + len,
+            sizeof(_buf) - start - sizeof(struct Header) - len);
```

Comment:
As I said, it's so similar that it has the same shortcomings of use a small buffers and using a memmove here :(.